File: app/Models/Booking.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use App\Models\Property;

class Booking extends Model
{
    protected $fillable = [
        'id_booking',
        'customer_id',
        'user_id',
        'bookable_id',
        'bookable_type',
        'assigned_room_id',
        'contact_name',
        'contact_email',
        'contact_phone',
        'institution',
        'total_price',
        'booking_type',
        'quantity',
        'status',
        'start_date',
        'end_date',
        'remember_token',
        'created_at',
        'updated_at',
    ];

    protected $hidden = [
        'remember_token',
    ];

    protected static function boot()
    {
        parent::boot();

        static::creating(function ($booking) {
            if (empty($booking->id_booking)) {
                $prefix = $booking->bookable_type === Property::class ? 'PRP' : 'TRN';
                $booking->id_booking = 'BKG-' . $prefix . '-' . date('Ymd') . '-' . strtoupper(substr(uniqid(), -5));
            }
        });
    }

    public function customer()

    {
        return $this->belongsTo(Customer::class);
    }

    public function assignedRoom()
    {
        return $this->belongsTo(Room::class, 'assigned_room_id');
    }

    public function bookable()
    {
        return $this->morphTo();
    }

    public function participants()
    {
        return $this->hasMany(Participant::class);
    }

    public function invoices()
    {
        return $this->hasMany(Invoice::class);
    }

    public function certificate()
    {
        return $this->hasOne(Certificate::class);
    }

    public function participantData()
    {
        return $this->hasMany(ParticipantData::class);
    }
}

File: frontend/lib/models/booking.dart
import 'package:frontend/models/property.dart';
import 'package:frontend/models/room.dart';
import 'package:frontend/models/user.dart';

class Booking {
  final int? id;
  final String? idBooking;
  final int? customerId;
  final int? propertyId;
  final String? contactName;
  final String? contactEmail;
  final String? contactPhone;
  final String? institution;
  final String? bookingType;
  final int? quantity;
  final String status;
  final DateTime startDate;
  final DateTime endDate;
  final Property? property;
  final Room? room;
  final User? user;

  Booking({
    this.id,
    this.idBooking,
    this.customerId,
    this.propertyId,
    this.contactName,
    this.contactEmail,
    this.contactPhone,
    this.institution,
    this.bookingType,
    this.quantity,
    required this.status,
    required this.startDate,
    required this.endDate,
    this.property,
    this.room,
    this.user,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'],
      idBooking: json['id_booking'],
      customerId: json['customer_id'],
      propertyId:
          json['property_id'] ??
          (json['property'] != null ? json['property']['id'] : null),
      contactName: json['contact_name'],
      contactEmail: json['contact_email'],
      contactPhone: json['contact_phone'],
      institution: json['institution'],
      bookingType: json['booking_type'],
      quantity: json['quantity'],
      status: json['status'],
      startDate: DateTime.parse(json['start_date']),
      endDate: DateTime.parse(json['end_date']),
      property: json['property'] != null
          ? Property.fromJson(json['property'])
          : null,
      room: json['room'] != null ? Room.fromJson(json['room']) : null,
      user: json['user'] != null ? User.fromJson(json['user']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'status': status,
      'start_date': startDate.toIso8601String(),
      'end_date': endDate.toIso8601String(),
    };
  }
}
